CONTRIBUTING.md

```diff
@@ -0,0 +1,226 @@
+command: `make -f misc/regen.mk clang-format-all`
+
+To summarize, the coding style resembles the Linux kernel's with the
+difference that it uses 4 spaces rather than one tab for indentation.
+
+Here's a `.vimrc` snippet that would use this style:
+
+```vim
+set tabstop=4
+set shiftwidth=4
+set softtabstop=4
+set expandtab
+```
+
+## Naming
+
+
+### Structs
+
+Structs are prefixed with `st_` and suffixed with `_t`. Structs that are
+found in public headers are typedef'ed and prefixed with `h2o_`, and
+suffixed with `_t`:
+
+```c
+/* public struct */
+typedef struct st_h2o_conn_t h2o_conn_t;
+
+struct st_h2o_conn_t {
+....
+};
+
+/* private struct */
+struct st_on_client_hello_ptls_t {
+...
+};
+```
+
+### Functions
+
+As with structs, publicly visible functions are prefixed with `h2o_`,
+such as `h2o_process_request`, whereas private functions don't use one:
+`get_ocsp_response`.
+
+### Goto labels
+
+Goto labels use upper case, and are indented with the matching code:
+
+```c
+void fn(const char *err)
+{
+    if (global) {
+        if (err == NULL)
+            goto Ok;
+
+    Ok:
+        this_is_ok();
+    }
+}
+```
+
+## Context passing and subclassing
+
+### Subclassing
+
+The H2O code base tends to use subclassing in order to pass contextual
+information. For example, both `h2o_http2_conn_t` and `struct
+st_h2o_http1_conn_t` are subclasses of `h2o_conn_t`. They do so by
+defining a `super` member of the struct of type `h2o_conn_t`, and
+returning that outside their respective modules. The struct is then cast
+back to the specilized subclasses when returning to their modules.
+
+This is in turn used so that both the HTTP/1 and HTTP/2 connections are
+able to expose a common interface for `get_sockname`, for example:
+
+```c
+/* http2 */
+static socklen_t get_sockname(h2o_conn_t *_conn, struct sockaddr *sa)
+{
+    h2o_http2_conn_t *conn = (void *)_conn;
+    ...
+}
+
+/* http1 */
+static socklen_t get_sockname(h2o_conn_t *_conn, struct sockaddr *sa)
+{
+    struct st_h2o_http1_conn_t *conn = (void *)_conn;
+    ...
+}
+```
+
+This way upper layers can act on an `h2o_conn_t` without exposing the
+details of `h2o_http2_conn_t` or `struct st_h2o_http1_conn_t`.
+
+### Context passing using struct member offsets
+
+Another technique commonly used in the H2O code base is offset
+calculations. This allows to better performance and reduced public interfaces, while
+still being able to pass objects of different types. H2O
+uses the `H2O_STRUCT_FROM_MEMBER` macro (which in turn uses
+[`offsetof`](https://en.wikipedia.org/wiki/Offsetof)) in order to
+compute the offset of a member in a struct. This way, given a pointer to
+a member, we can obtain a pointer to the enclosing struct.
+
+Linked lists (`h2o_linklist_t`) and timers (`h2o_timeout_entry_t`)
+are typical users of the technique.
+
+Here's an example demonstrating how to pass context alongside a timer pointer:
+```c
+struct st_mycontext_t {
+  void *ctx;
+  h2o_timeout_entry_t timer;
+};
```

Comment:
How about using an example that embeds the context within the struct, instead of using one that uses a pointer to refer to a value external from the object?

The performance benefit of using `H2O_STRUCT_FROM_MEMBER` is that it reduces indirection (i.e. pointer deference). It is true that we sometimes need to refer to information using pointers, I think that using _that_ as an example might give the readers a false notion that it is the suggested way.